File: internal/rest/sse.go
package rest

import (
	"anova4all/pkg/wifi"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"time"

	"github.com/google/uuid"
	"github.com/puzpuzpuz/xsync"
)

type SSEManager struct {
	deviceManager wifi.AnovaManager
	listeners     *xsync.MapOf[string, *xsync.MapOf[string, chan SSEEvent]]
}

func NewSSEManager(deviceManager wifi.AnovaManager) *SSEManager {
	manager := &SSEManager{
		deviceManager: deviceManager,
		listeners:     xsync.NewMapOf[*xsync.MapOf[string, chan SSEEvent]](),
	}
	manager.registerCallbacks()
	return manager
}

func (m *SSEManager) Connect(deviceID string) (string, <-chan SSEEvent) {
	listenerID := uuid.NewString()
	eventChan := make(chan SSEEvent, 100)

	deviceListeners, _ := m.listeners.LoadOrStore(deviceID, xsync.NewMapOf[chan SSEEvent]())
	deviceListeners.Store(listenerID, eventChan)

	return listenerID, eventChan
}

func (m *SSEManager) Disconnect(deviceID, listenerID string) {
	if deviceListeners, ok := m.listeners.Load(deviceID); ok {
		if ch, loaded := deviceListeners.LoadAndDelete(listenerID); loaded {
			close(ch)
		}
		if deviceListeners.Size() == 0 {
			m.listeners.Delete(deviceID)
		}
	}
}

func (m *SSEManager) Broadcast(event SSEEvent) {
	if deviceListeners, ok := m.listeners.Load(event.DeviceID); ok {
		deviceListeners.Range(func(_ string, ch chan SSEEvent) bool {
			go func(ch chan SSEEvent) {
				ch <- event
			}(ch)
			return true
		})
	}
}

func (m *SSEManager) DeviceConnectedCallback(device wifi.AnovaDevice) {
	event := SSEEvent{
		DeviceID:  device.IDCard(),
		EventType: SSEEventTypeDeviceConnected,
		Timestamp: time.Now(),
	}
	m.Broadcast(event)
}

func (m *SSEManager) DeviceDisconnectedCallback(deviceID string) {
	event := SSEEvent{
		DeviceID:  deviceID,
		EventType: SSEEventTypeDeviceDisconnected,
		Timestamp: time.Now(),
	}
	m.Broadcast(event)
}

func (m *SSEManager) DeviceStateChangeCallback(deviceID string, state wifi.DeviceState) {
	event := SSEEvent{
		DeviceID:  deviceID,
		EventType: SSEEventTypeStateChanged,
		Payload:   state,
		Timestamp: time.Now(),
	}
	m.Broadcast(event)
}

func (m *SSEManager) DeviceEventCallback(deviceID string, anovaEvent wifi.AnovaEvent) {
	event := SSEEvent{
		DeviceID:  deviceID,
		EventType: SSEEventTypeEvent,
		Payload:   anovaEvent,
		Timestamp: time.Now(),
	}
	m.Broadcast(event)
}

func (m *SSEManager) registerCallbacks() {
	m.deviceManager.OnDeviceConnected(m.DeviceConnectedCallback)
	m.deviceManager.OnDeviceDisconnected("*", m.DeviceDisconnectedCallback)
	m.deviceManager.OnDeviceStateChange("*", m.DeviceStateChangeCallback)
	m.deviceManager.OnDeviceEvent("*", m.DeviceEventCallback)
}

func (m *SSEManager) ServeHTTP(deviceID string) func(w http.ResponseWriter, r *http.Request) {
	return func(w http.ResponseWriter, r *http.Request) {
		listenerID, eventChan := m.Connect(deviceID)
		defer m.Disconnect(deviceID, listenerID)

		ctx, cancel := context.WithCancel(r.Context())
		defer cancel()

		w.Header().Set("Content-Type", "text/event-stream")
		w.Header().Set("Cache-Control", "no-cache")
		w.Header().Set("Connection", "keep-alive")
		w.WriteHeader(http.StatusOK)

		flusher, ok := w.(http.Flusher)
		if !ok {
			http.Error(w, "Streaming unsupported", http.StatusInternalServerError)
			return
		}

		pingTicker := time.NewTicker(30 * time.Second)
		defer pingTicker.Stop()

		for {
			select {
			case <-ctx.Done():
				return
			case event := <-eventChan:
				data, _ := json.Marshal(event)
				fmt.Fprintf(w, "event: %s\ndata: %s\n\n", event.EventType, data)
				flusher.Flush()
			case <-pingTicker.C:
				fmt.Fprintf(w, "event: ping\ndata: {}\n\n")
				flusher.Flush()
			}
		}
	}
}

File: python/app/sse.py
import asyncio
import uuid
from typing import Dict, AsyncIterator

from pydantic import BaseModel

from anova_wifi.device import AnovaDevice, DeviceState
from anova_wifi.event import AnovaEvent
from anova_wifi.manager import AnovaManager
from .models import SSEEvent, SSEEventType


async def event_stream(resp: AsyncIterator[BaseModel]) -> AsyncIterator[str]:
    async for event in resp:
        evnt = event.__class__.__name__
        if hasattr(event, "event_type"):
            evnt = event.event_type
        yield f"event: {evnt}\ndata: {event.model_dump_json()}\n\n"


class SSEManager:
    _listeners: Dict[str, Dict[str, asyncio.Queue[SSEEvent]]] = {}

    def __init__(self, device_manager: AnovaManager):
        self.device_manager = device_manager

    async def connect(self, device_id: str) -> tuple[str, asyncio.Queue[SSEEvent]]:
        if device_id not in self._listeners:
            self._listeners[device_id] = {}

        listener_id = str(uuid.uuid4())
        queue: asyncio.Queue[SSEEvent] = asyncio.Queue()
        self._listeners[device_id][listener_id] = queue
        return listener_id, queue

    async def disconnect(self, device_id: str, listener_id: str) -> None:
        if device_id in self._listeners and listener_id in self._listeners[device_id]:
            del self._listeners[device_id][listener_id]
            if not self._listeners[device_id]:
                del self._listeners[device_id]

    async def broadcast(self, event: SSEEvent) -> None:
        device_id = event.device_id
        if device_id in self._listeners:
            for queue in self._listeners[device_id].values():
                await queue.put(event)

    async def device_connected_callback(self, device: AnovaDevice) -> None:
        event = SSEEvent(
            device_id=device.id_card,
            event_type=SSEEventType.device_connected,
        )
        await self.broadcast(event)

    async def device_disconnected_callback(self, device_id: str) -> None:
        event = SSEEvent(
            device_id=device_id,
            event_type=SSEEventType.device_disconnected,
        )
        await self.broadcast(event)

    async def device_state_change_callback(self, device_id: str, state: DeviceState) -> None:
        event = SSEEvent(
            device_id=device_id,
            event_type=SSEEventType.state_changed,
            payload=state
        )
        await self.broadcast(event)

    async def device_event_callback(self, device_id: str, event: AnovaEvent) -> None:
        await self.broadcast(SSEEvent(
            device_id=device_id,
            event_type=SSEEventType.event,
            payload=event
        ))

    def register_callbacks(self) -> None:
        self.device_manager.on_device_connected(self.device_connected_callback)
        self.device_manager.on_device_disconnected("*", self.device_disconnected_callback)
        self.device_manager.on_device_state_change("*", self.device_state_change_callback)
        self.device_manager.on_device_event("*", self.device_event_callback)
